src/formats/maeformat.cpp

```diff
@@ -0,0 +1,333 @@
+#include <openbabel/atom.h>
+#include <openbabel/bond.h>
+#include <openbabel/obiter.h>
+#include <openbabel/elements.h>
+#include <openbabel/generic.h>
+
+#include <iostream>
+#include <map>
+
+#include <MaeConstants.hpp>
+#include <Reader.hpp>
+#include <Writer.hpp>
+
+using namespace std;
+using namespace schrodinger::mae;
+using boost::dynamic_bitset;
+namespace OpenBabel
+{
+
+class MAEFormat : public OBMoleculeFormat
+{
+public:
+  //Register this format type ID in the constructor
+  MAEFormat()
+	{
+		OBConversion::RegisterFormat("MAE", this);
+		OBConversion::RegisterFormat("MAEGZ", this);
+	}
+
+	virtual const char* Description() //required
+	{
+		return
+		"Maestro format\n"
+		"File format of Schrödinger Software\n";
+    };
+
+    //URL where the file format is specified
+    virtual const char* SpecificationURL()
+    {
+        return "https://github.com/schrodinger/maeparser";
+    };
+
+    virtual int SkipObjects(int n, OBConversion* pConv);
+
+    ////////////////////////////////////////////////////
+    /// Declarations for the "API" interface functions. Definitions are below
+    virtual bool ReadMolecule(OBBase* pOb, OBConversion* pConv);
+    virtual bool WriteMolecule(OBBase* pOb, OBConversion* pConv);
+
+private:
+    // TODO:  This will be implemented in maeparser more completely,
+    // we should migrate to that when available.
+    const map<int, int> atomic_num_to_color = {
+        {1, 21}, {6, 2}, {7, 43}, {8, 70}, {9, 8}, {16, 13}, {17, 9} };
+
+    shared_ptr<Block> m_next_mae;
+    shared_ptr<Reader> m_reader;
+    shared_ptr<Writer> m_writer;
+
+    shared_ptr<IndexedBlock> TranslateAtomBlock(OBMol* pmol);
+    shared_ptr<IndexedBlock> TranslateBondBlock(OBMol* pmol);
+
+    void setupReader(OBConversion* pConv);
+    void checkEOF(OBConversion* pConv);
+
+};
+	////////////////////////////////////////////////////
+
+//Make an instance of the format class
+MAEFormat theMAEFormat;
+
+/////////////////////////////////////////////////////////////////
+
+int MAEFormat::SkipObjects(int n, OBConversion* pConv)
+{
+    // Required for the MaeParser interface, create a shared_ptr w/o
+    // memory management
+    setupReader(pConv);
+    for(int i=0; i<n; i++) {
+        m_next_mae = m_reader->next(CT_BLOCK);
+        checkEOF(pConv);
+        if(m_next_mae==nullptr) {
+            return 0;
+        }
+    }
+    return 0;
+};
+
+
+void MAEFormat::setupReader(OBConversion* pConv)
+{
+    static string filename = "";
+    if(filename == pConv->GetInFilename()) return;
+    filename = pConv->GetInFilename();
+
+    // Required for the MaeParser interface, create a shared_ptr w/o
+    // memory management
+    shared_ptr<istream> ifs(shared_ptr<istream>(), pConv->GetInStream());
+    m_reader = make_shared<Reader>(ifs);
+
+    m_next_mae = m_reader->next(CT_BLOCK);
+}
+
+
+/* Guilt disclaimer:  This is an ugly hack, but I think required.
+ *
+ * Since maeparser buffers what it reads, we have to do some cheating
```

Comment:
There originally was a non-buffered mode of maeparser, and it may still work, but the truth is it's a lot slower and less tested, and this works, so I think sticking with this hack makes sense.